File: client/primes.py
#!/usr/bin/env python3
import math
from gmpy2 import powmod
from random import randint
import time

def isDiv(num, div):
    return (num // div) * div == num

def isOdd(num):
    return (num % 2 == 1)

def twoSD(num):
    i = 1
    D = (num-1) // 2
    while not isOdd(D):
        D //= 2
        i+=1

    S = i
    ##print(num, (2**S)*D+1)
    #print(num, S, D)
    return (int(S), int(D))

def sieveCheck(num, trials=25):
    primeTable = [2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41, 43, 47, 53, 59, 61, 67, 71, 73, 79, 83, 89, 97, 101, 103, 107, 109, 113, 127, 131, 137, 139, 149, 151, 157, 163, 167, 173, 179, 181, 191, 193, 197, 199, 211, 223, 227, 229, 233, 239, 241, 251, 257, 263, 269, 271, 277, 281, 283, 293, 307, 311, 313, 317, 331, 337, 347, 349, 353, 359, 367, 373, 379, 383, 389, 397, 401, 409, 419, 421, 431, 433, 439, 443, 449, 457, 461, 463, 467, 479, 487, 491, 499, 503, 509, 521, 523, 541, 547, 557, 563, 569, 571, 577, 587, 593, 599, 601, 607, 613, 617, 619, 631, 641, 643, 647, 653, 659, 661, 673, 677, 683, 691, 701, 709, 719, 727, 733, 739, 743, 751, 757, 761, 769, 773, 787, 797, 809, 811, 821, 823, 827, 829, 839, 853, 857, 859, 863, 877, 881, 883, 887, 907, 911, 919, 929, 937, 941, 947, 953, 967, 971, 977, 983, 991, 997]
    for i in range(0, trials):
        if isDiv(num, primeTable[i]):
            return False
    return True

def isPrime(num, iterations=64, trials=25):
    if not sieveCheck(num, trials=25):
        return False
    for i in range(0, iterations):
        #print(i)
        a = randint(2, num - 2) # generate A
        num_rewrite = twoSD(num)
        y = powmod(a, num_rewrite[1], num)
        #print(y)
        if y != 1 and y != num-1:
            j = 1
            while j <= (num_rewrite[0]-1) and y != num-1:
                #print(j)
                y = pow(y, 2, num)
                #print(y)
                if y == 1:
                    #if i != 0: print(i)
                    return False
                j = j+1
            if y != num-1:
                #if i != 0: print(i)
                return False
    #print(i)
    return True

def primeGen(bits=256, iterations=128, trials=25):
    #print(twoSD(48112959837082048697))
    found = False
    while not found:
        prime_candidate = randint(3, pow(2, bits-1))*2-1
        found = isPrime(prime_candidate, trials=25)
        #print("uh")
    return prime_candidate

#print(primeGen(2048, iterations = 256, trials = 150))

File: client/primes/primes.c
#include <stdio.h>
#include <gmp.h>
#include <stdbool.h>
#include <time.h>

bool isDiv(mpz_t num, unsigned int div) {
	mpz_t temp_algebra;
	mpz_init(temp_algebra);
	mpz_tdiv_r_ui(temp_algebra, num, div);
	bool returnval = (temp_algebra == 0);
	//bool returnval = (num / div) * div == num;
	mpz_clear(temp_algebra);
	return returnval;
}

bool isOdd(mpz_t num) {
	return mpz_tdiv_r_ui(num, num, 2) == 1;
	//bool returnval = (num % 2 == 1);
	//return returnval;
}

void twoSD(mpz_t num, int *S_ptr, mpz_t *D_ptr, mpz_t *numM1_ptr) {
	mpz_sub_ui(*numM1_ptr, num, 1);
	*S_ptr = 0;
	mpz_tdiv_q_ui(*D_ptr, *numM1_ptr, 2);
	//*D_ptr = (num-1 / 2);
	while (!isOdd(*D_ptr)) {
		mpz_tdiv_q_ui(*D_ptr, *D_ptr, 2);
		//*D_ptr = *D_ptr / 2;
		*S_ptr++;
	}
}

bool sieveCheck(mpz_t num) {
	int i;
	int firstPrimes[] = {2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41, 43, 47, 53, 59, 61, 67, 71, 73, 79, 83, 89, 97};
	for (i = 0; i < 25; i++) {
		if (isDiv(num, firstPrimes[i])) {
			return false;
		}
	}
	return true;
}

bool isPrime(mpz_t num, int iterations) {
	mpz_t D; // (longun)
	mpz_t y; // (longun)
	mpz_t numM1;
	mpz_init(numM1);
	mpz_init(D);
	mpz_init(y);
	int S;
	int j;

	mpz_t a;		// sorting out random numbers
	mpz_init(a);		// this will be the one written to
	
	gmp_randstate_t state;	// so this is the state (?)
	gmp_randinit_default(state); // setting the state (normal generator)
	gmp_randseed_ui(state, (unsigned long)time(NULL)); // better numbers

	if (!sieveCheck(num))
		return false;
	for (int i=0; i<iterations; i++) {
		mpz_urandomb(a, state, 2048); // this loads the actual rand #
		twoSD(num, &S, &D, &numM1);
		mpz_powm(y, a, D, num);
		// mod exp (a, D, num)
		if (!mpz_cmp_si(y, 1) && !mpz_cmp(y, numM1)) {
			j = 1;
			while (j < S && !mpz_cmp(y, numM1)) {
				mpz_powm_ui(y, y, 2, num);
				// y = y^2 mod num
				if (mpz_cmp_si(y, 1)) {
					mpz_clear(D);
					mpz_clear(y);
					mpz_clear(a);
					mpz_clear(numM1);
					gmp_randclear(state);
					return false;
				}
				j++;
			}
			if (!mpz_cmp(y, numM1)) {
				mpz_clear(D);
				mpz_clear(y);
				mpz_clear(a);
				mpz_clear(numM1);
				gmp_randclear(state);
				return false;
			}
		}
	}
	mpz_clear(D);
	mpz_clear(y);
	mpz_clear(a);
	mpz_clear(numM1);
	gmp_randclear(state);
	return true;
}

void primeGen(int bits, int iterations) {
	bool found = false;
	mpz_t candidate;	// sorting out random numbers
	mpz_init(candidate);	// this will be the one written to
	
	gmp_randstate_t cand_state;	// so this is the state (?)
	gmp_randinit_default(cand_state); // setting the state (normal generator)
	gmp_randseed_ui(cand_state, (unsigned long)time(NULL)); // better numbers
	while (!found) {
		mpz_urandomb(candidate, cand_state, bits); // this loads the actual rand #
		//candidate = 5; // longun, random between 3 and 2^bits
		found = isPrime(candidate, iterations);
		puts("failure.\n");
	}
	gmp_printf("%Zd\n", candidate);
	mpz_clear(candidate);
	gmp_randclear(cand_state);
}

int main() {
	primeGen(128, 64);
	//bool x = isDiv(15, 5);
	//printf("%B\n", x);
	return 0;
}
